### PR TITLE
NODE-2524: Correct Type Checking Errors

### PR DESCRIPTION
**Pull request by nbbeeken**
## Correct Type Errors

 I went through and fixed as best I could by adding good type information, but sometimes `any` had to be used. There's documentation that needs filling in that wasn't there prior all marked with `TODO:(ts)` I'll ask for advice on how to find the information and best document it.

### CODE REVIEW

File: lib/connection_string.js [hunk 1]
```diff
@@ -669,7 +674,7 @@ function parseConnectionString(uri, options, callback) {
         parsedHost.port = null;
       }
 
-      if (Number.isNaN(parsedHost.port)) {
+      if (Number.isNaN(+parsedHost.port)) {
```
Comment (reggi) [inline lib/connection_string.js:677]: I've never really used `NaN` type guards, what's the idea of `Number.isNaN(+` here?

Comment (nbbeeken) [reply lib/connection_string.js:677]: `+` in front of a variable coerces it to be a number even `+'NaN' === NaN`, typescript complains cus port is a string.

File: lib/sdam/server_description.js [hunk 2]
```diff
@@ -66,6 +69,28 @@ class ServerDescription {
       ismaster
     );
 
+    // https://github.com/mongodb/specifications/blob/master/source/server-discovery-and-monitoring/server-discovery-and-monitoring.rst#serverdescription
+    // set up defaults / required properties
+    this.address = address;
+    this.error = null;
+    this.roundTripTime = null;
```
Comment (emadum) [inline lib/sdam/server_description.js:76]: Some of these get immediately re-assigned a few lines down, doesn't look intended.

File: lib/operations/db_ops.js [hunk 3]
```diff
@@ -203,12 +206,12 @@ function createIndex(db, name, fieldOrSpec, options, callback) {
...
-      err.code === 85 ||
-      err.code === 86 ||
-      err.code === 11600 ||
-      err.code === 197
+      /** @type {Error & {code?:number}} */ (err).code === 67 ||
+      /** @type {Error & {code?:number}} */ (err).code === 11000 ||
+      /** @type {Error & {code?:number}} */ (err).code === 85 ||
+      /** @type {Error & {code?:number}} */ (err).code === 86 ||
+      /** @type {Error & {code?:number}} */ (err).code === 11600 ||
+      /** @type {Error & {code?:number}} */ (err).code === 197
```
Comment (emadum) [inline lib/operations/db_ops.js:214]: I'm guessing this pattern is necessary to fix a specific error; it seems pretty awkward. Is there anything else we could do here, possibly an ignore?

Comment (nbbeeken) [reply lib/operations/db_ops.js:214]: I could do `err['code']` or `//@ts-ignore` on each line .code doesn't exist on the type that err is because I believe err here is either a MongoError, Error, or a (error) result object from the db and its only the last one that has `.code` open to whatever we think is best

File: lib/sdam/server_description.js (hunk 2 above)
Comment (nbbeeken) [reply lib/sdam/server_description.js:76]: I guess I left it like that because I wanted to make it clear that these were the defaults as specified in the spec and that I had to add them in order to get tsc to know that they are indeed properties of the class. But it is indeed clunky and potentially confusing so I'll combine and reduce

File: lib/bulk/common.js [hunk 4]
```diff
@@ -1130,9 +1134,9 @@ class BulkOperationBase {
       delete finalOptions.bypassDocumentValidation;
     }
 
-    // Set an operationIf if provided
-    if (this.operationId) {
-      config.resultHandler.operationId = this.operationId;
+    // Set an operationId if provided
+    if (this['operationId']) {
+      config.resultHandler['operationId'] = this['operationId'];
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply lib/bulk/common.js:1139]: The operationId property doesn't exist on this class it exists on the subclasses, I can do this to make the type error go away or add `this.operationId = undefined` in the constructor.

The other places I did it were for example, options objects where we add some property on to the options before sending it deeper into the driver, since the prop hadn't existed until that point the prop didn't exist on the type of the object, I could just call the object `any` but there was some typing already in place that I didn't want to lose. Future work that I suppose I should make a ticket for would be to reduce `any`'s and `obj['blah']` as part of increasing our typing strength

File: lib/bulk/common.js [hunk 5]
```diff
@@ -1,8 +1,7 @@
 'use strict';
 
-const {
-  BSON: { Long, ObjectId }
-} = require('../deps');
+const { BSON } = require('../deps');
+const { ObjectId, Long } = BSON;
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply lib/bulk/common.js:4]: So this I broke out for an opinionated reason and totally can revert, but @mbroadst is working on fixing the bug that is blocking us from generating type definitions from all this jsdoc however it may not be fixable in the near future.

Since an alternative would be to change up the way we import and export things I investigated that es6 module `import`/`export` syntax particularly `import { BSON } from '../deps';` while it looks like object de-structuring it isn't related at all apparently so it does not support `import { BSON: {ObjectId, Long} } from '../deps';`, this is me jumping the gun and cleaning up that behavior in advance.
Also I don't love that it breaks it across two lines.
Like I said TOTALLY opinionated but with some reasoning behind it 😅

File: .eslintrc [hunk 6]
```diff
@@ -29,7 +29,8 @@
 
     "no-console": "off",
     "eqeqeq": ["error", "always", { "null": "ignore" }],
-    "strict": ["error", "global"]
+    "strict": ["error", "global"],
+    "no-unused-vars": ["error", { "argsIgnorePattern": "^_" }]
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply .eslintrc:33]: Good catch, I'm ok with double underscore, lemme loop in @mbroadst and @reggi on this

File: lib/operations/db_ops.js (hunk 3 above)
Comment (emadum) [reply lib/operations/db_ops.js:214]: How about
```
    const code = /** @type {Error & {code?:number}} */ (err).code;
    if (
      code === 67 ||
      code === 11000 ||
      code === 85 ||
      code === 86 ||
      code === 11600 ||
      code === 197
    ) {
      return handleCallback(callback, err, result);
    }
```

File: lib/bulk/common.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (emadum) [reply lib/bulk/common.js:1139]: Hmm, I think I'd prefer just having a `// @ts-ignore` on those lines. In actual TS we'd probably use an abstract class for something like this in order to get good checks, but I'd rather add comments than tweak the code in cases where we're not actually getting any benefit from the checks.

[the remark this answers is not in the document]
Comment (nbbeeken) [reply lib/bulk/common.js]: I'm inclined to agree, it also marks the issues for future fixing with a consistent comment, cus we also use `obj['prop']` notation elsewhere on purpose

File: lib/operations/db_ops.js (hunk 3 above)
Comment (nbbeeken) [reply lib/operations/db_ops.js]: Much cleaner, went with that 😃

File: lib/sdam/topology.js [hunk 7]
```diff
@@ -101,16 +112,34 @@ class Topology extends EventEmitter {
    * @param {number} [options.localThresholdMS=15] The size of the latency window for selecting among multiple suitable servers
    * @param {number} [options.serverSelectionTimeoutMS=30000] How long to block for server selection before throwing an error
    * @param {number} [options.heartbeatFrequencyMS=10000] The frequency with which topology updates are scheduled
+   * @param {MongoCredentials} [options.credentials] Authentication credentials.
+   * @param {any} [options.cursorFactory] a Cursor constructor class.
```
Comment (mbroadst) [inline lib/sdam/topology.js:116]: Why are we using `any` for all of these types? Was it just exhaustion at dealing with the codebase? I think a number of them are pretty explicitly some other type (`cursorFactory` is probably a `CoreCursor`, etc).

File: lib/sdam/topology.js [hunk 8]
```diff
@@ -213,7 +241,7 @@ class Topology extends EventEmitter {
   }
 
   /**
-   * @returns A `TopologyDescription` for this topology
+   * @returns {TopologyDescription} for this topology
```
Comment (mbroadst) [inline lib/sdam/topology.js:244]: this doc string is weird now

File: lib/sdam/topology.js [hunk 9]
```diff
@@ -223,7 +251,8 @@ class Topology extends EventEmitter {
    * Initiate server connect
    *
    * @param {object} [options] Optional settings
-   * @param {Array} [options.auth=null] Array of auth options to apply on connect
+   * @param {Array} [options.auth] Array of auth options to apply on connect
+   * @param {any} [options.readPreference] TODO:(ts) types and documentation
```
Comment (mbroadst) [inline lib/sdam/topology.js:255]: I'll stop comment on these now, but remember this goes right into _user-facing_ documentation. Imagine seeing that

File: lib/sdam/topology.js [hunk 10]
```diff
@@ -582,7 +613,7 @@ class Topology extends EventEmitter {
    * @param {boolean} [options.ignoreUndefined=false] Specify if the BSON serializer should ignore undefined fields
    * @param {ClientSession} [options.session] Session to use for the operation
    * @param {boolean} [options.retryWrites] Enable retryable writes for this operation
-   * @param {opResultCallback} callback A callback function
+   * @param {(error?: Error, result?: any) => void} callback A callback function
```
Comment (mbroadst) [inline lib/sdam/topology.js:616]: typedef the callback type?

File: lib/sdam/topology.js [hunk 11]
```diff
@@ -698,6 +734,7 @@ class Topology extends EventEmitter {
    * @param {boolean} [options.ignoreUndefined=false] Specify if the BSON serializer should ignore undefined fields.
    * @param {ClientSession} [options.session=null] Session to use for the operation
    * @param {object} [options.topology] The internal topology of the created cursor
+   * @param {{new(topology: any, ns: string | MongoDBNamespace, cmd: object, options: object)}} [options.cursorFactory] A Cursor constructor
```
Comment (mbroadst) [inline lib/sdam/topology.js:737]: this looks like it needs a typedef, I think you might be able to just put `CoreCursor` here

File: lib/sdam/topology.js [hunk 12]
```diff
@@ -1088,6 +1123,8 @@ const isRetryableWritesSupported = function(topology) {
   return true;
 };
 
+module.exports = { Topology };
```
Comment (mbroadst) [inline lib/sdam/topology.js:1126]: same question as above, why is the exports moved up here?

File: lib/sessions.js [hunk 13]
```diff
@@ -549,7 +556,7 @@ class ServerSession {
       (((Date.now() - this.lastUse) % 86400000) % 3600000) / 60000
     );
 
-    return idleTimeMinutes > sessionTimeoutMinutes - 1;
+    return idleTimeMinutes > +sessionTimeoutMinutes - 1;
```
Comment (mbroadst) [inline lib/sessions.js:559]: this might just mean that `sessionTimeoutMinutes` is incorrectly documented above. Perhaps it should just be `Number`

File: lib/utils.js [hunk 14]
```diff
@@ -644,23 +659,14 @@ function deprecateOptions(config, fn) {
...
-
 class MongoDBNamespace {
   constructor(db, collection) {
     this.db = db;
     this.collection = collection;
   }
 
   toString() {
-    return this.collection ? `${this.db}.${this.collection}` : this.db;
+    return this.collection ? `${this.db}.${this.collection}` : `${this.db}`;
```
Comment (mbroadst) [inline lib/utils.js:669]: is this change required because `this.db` doesn't have type information? Would it still be required if jsdoc's were added for this type with correct type information?

File: lib/utils.js [hunk 15]
```diff
@@ -769,7 +775,7 @@ function relayEvents(listener, emitter, events) {
  * instances
  *
  * @private
- * @param {(Topology|Server)} topologyOrServer
+ * @param {Topology | Server | any} topologyOrServer TODO:(ts) better types
```
Comment (mbroadst) [inline lib/utils.js:778]: where is `any` passed in? I think that's probably a bug.

File: .eslintrc (hunk 6 above)
[the remark this answers is not in the document]
Comment (nbbeeken) [reply .eslintrc:33]: Typescript complains about super classes that have methods that lack the parameters being used by calls to the said function. eslint-jdoc plugin also complains about documenting params that don't exist. A good example is in auth_providers.js.
the jdoc rule I suppose could be disabled but typescript we would need to keep //@ts-ignores in any super class that uses an abstract function.

File: index.js [hunk 16]
```diff
@@ -1,65 +1,75 @@
...
+  Map: BSON['Map'], // TODO:(neal) should this just be Map?
+  DBRef: BSON.DBRef,
+  Double: BSON.Double,
+  Int32: BSON.Int32,
+  Long: BSON.Long,
+  MinKey: BSON.MinKey,
+  MaxKey: BSON.MaxKey,
+  ObjectId: BSON.ObjectId,
+  ObjectID: BSON.ObjectId,
+  BSONSymbol: BSON['BSONSymbol'], // TODO:(neal) this is missing from bson types?
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply index.js:68]: It's missing from the community [@types/bson](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/bson/index.d.ts#L430) which I was leaning on for a bit while fixing types, I'll change it back to dot op access as per my conversation with Eric about undoing `obj['prop']` that I added

File: lib/bulk/common.js [hunk 17]
```diff
@@ -923,8 +927,8 @@ class BulkOperationBase {
    * Specifies a raw operation to perform in the bulk write.
    *
    * @function
-   * @param {object} op The raw operation to perform.
-   * @param {boolean} [op.hint] An optional hint for query optimization. See the {@link https://docs.mongodb.com/manual/reference/command/update/#update-command-hint|update command} reference for more information.
+   * @param {any} op The raw operation to perform.
+   * //param {boolean} [op.hint] An optional hint for query optimization. See the {@link https://docs.mongodb.com/manual/reference/command/update/#update-command-hint|update command} reference for more information.
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply lib/bulk/common.js:931]: I should mark this as a TODO so it doesn't get lost but op is a complicated object basically, so If I leave it as `any` then I can't define `[op.hint]` on `any`. The alternative to typing out the object was just unclear as the structure changes based on which op it contains, e.g. op.insertMany is an array but the others aren't, if we want to strongly type it its gunna be an interesting conditional type.
`any` seemed acceptable for this iteration but I didn't wanna lose the hint documentation
Ofc there's probably a better way to do this for these tricky types

File: lib/change_stream.js [hunk 18]
```diff
@@ -19,19 +25,19 @@ const CHANGE_DOMAIN_TYPES = {
...
- * @typedef OperationTime
+ * @typedef {import('bson').Timestamp} OperationTime
  * @description Represents a specific point in time on a server. Can be retrieved by using {@link Db#command}
  * @see https://docs.mongodb.com/manual/reference/method/db.runCommand/#response
  */
 
 /**
- * @typedef ChangeStreamOptions
- * @description Options that can be passed to a ChangeStream. Note that startAfter, resumeAfter, and startAtOperationTime are all mutually exclusive, and the server will error if more than one is specified.
+ * @typedef ChangeStreamOptions Options that can be passed to a ChangeStream. Note that startAfter, resumeAfter, and startAtOperationTime are all mutually exclusive, and the server will error if more than one is specified.
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply lib/change_stream.js:40]: JSDoc plugin for eslint demands typedef be followed immediately by `@property` (if any exist) I realize now I can move description to the end of the property list and it will be happy. However, I see that VSCode only picks up the description for hover docs if the description follows the typedef like this, that certainly doesn't have to be the rule we live by though

File: lib/cmap/auth/auth_provider.js [hunk 19]
```diff
@@ -80,8 +84,13 @@ class AuthProvider {
...
    * Implementation of a single connection authenticating. Is meant to be overridden.
    * Will error if called directly
+   *
+   * @param {any} _sendAuthCommand
+   * @param {any} _connection
+   * @param {any} _credentials
+   * @param {any} _callback
    */
-  _authenticateSingleConnection(/*sendAuthCommand, connection, credentials, callback*/) {
+  _authenticateSingleConnection(_sendAuthCommand, _connection, _credentials, _callback) {
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply lib/cmap/auth/auth_provider.js:93]: Answered in the convo where I AT-ed the team.

File: lib/cmap/connection_pool.js [hunk 20]
```diff
@@ -493,10 +497,14 @@ function processWaitQueue(pool) {
   }
 }
 
+module.exports = {
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply lib/cmap/connection_pool.js:500]: Ah annoying thing, I can move it back but VSCode, (and I think in turn tsc, but I have to check) was picking up the event documentation as a type annotation for ConnectionPool, so it became an event and not the class it is declared to be.

File: lib/deps.js [hunk 21]
```diff
@@ -17,8 +17,10 @@ function makeErrorModule(error) {
   });
 }
 
+/** @type {import('bson')} */
 let BSON = require('bson');
 try {
+  // @ts-ignore NOTE: optional dependency
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply lib/deps.js:23]: Unfortunately tsc has a checker for module existence, so even with a typecast the error remains.

File: lib/gridfs-stream/download.js [hunk 22]
```diff
@@ -1,164 +1,151 @@
...
  * @class
- * @extends external:Readable
+ * @extends Readable
  * @param {Collection} chunks Handle for chunks collection
  * @param {Collection} files Handle for files collection
  * @param {object} readPreference The read preference to use
  * @param {object} filter The query to use to find the file document
  * @param {object} [options] Optional settings.
- * @param {number} [options.sort] Optional sort for the file find query
+ * @param {{uploadDate?: number, [sortKey: string]: number}} [options.sort] Optional sort for the file find query
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply lib/gridfs-stream/download.js:22]: Yea probably a good idea, spec says that sort is an optional doc so maybe actually the opposite, should go more generic?

https://github.com/mongodb/specifications/blob/master/source/gridfs/gridfs-spec.rst#generic-find-on-files-collection

File: lib/operations/connect.js [hunk 23]
```diff
@@ -299,12 +298,8 @@ function createTopology(mongoClient, options, callback) {
   // Pass in the promise library
   options.promiseLibrary = mongoClient.s.promiseLibrary;
 
-  const translationOptions = {
-    createServers: false
-  };
-
   // Set default options
-  translateOptions(options, translationOptions);
+  translateOptions(options);
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply lib/operations/connect.js:302]: https://github.com/mongodb/node-mongodb-native/blob/a244966dc624371bed4a4b13530b30ad843bab16/lib/operations/connect.js#L547
It doesn't seem to take a second argument as of 
https://github.com/mongodb/node-mongodb-native/commit/6aa2434628e85ead8e5be620c27ebe8ab08a1c05#diff-2bafce404f7924755458c0320d824b53R573

File: lib/operations/operation.js [hunk 24]
```diff
@@ -41,7 +43,10 @@ class OperationBase {
     return true;
   }
 
-  execute() {
+  /**
+   * @param {Function} _callback
+   */
+  execute(_callback) {
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply lib/operations/operation.js:49]: Unfortunately it is a tsc error that needs to be ignored at the call points of `execute`, if that's acceptable I can go that way.

File: lib/sdam/server_description.js [hunk 25]
```diff
@@ -38,6 +38,8 @@ const ISMASTER_FIELDS = [
   '$clusterTime'
 ];
 
+/** @typedef {'minWireVersion' | 'maxWireVersion' | 'maxBsonObjectSize' | 'maxMessageSizeBytes' | 'maxWriteBatchSize' | 'compression' | 'me' | 'hosts' | 'passives' | 'arbiters' | 'tags' | 'setName' | 'setVersion' | 'electionId' | 'primary' | 'logicalSessionTimeoutMinutes' | 'saslSupportedMechs' | '__nodejs_mock_server__' | '$clusterTime'} IsMasterProperties */
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply lib/sdam/server_description.js:41]: Good catch will remove.
I was trying to do something clever to avoid all the default `this.x =` initializers I had to add, if it were .ts you can do this thing with an array of strings turn it into a union and then with an annotation tsc would know that these props exist on ServerDescription, without actually declaring them.

File: lib/sdam/topology.js (hunk 7 above)
Comment (nbbeeken) [reply lib/sdam/topology.js:116]: Because of that exports issue, `CoreCursor` is not constructable. Yes exhaustion factored in too 🙏 but this unfortunately is just a type error in the current state
I could give it a better type annotation like here: (maybe hoist it to a typedef as per your comment for this line)
https://github.com/mongodb/node-mongodb-native/blob/a244966dc624371bed4a4b13530b30ad843bab16/lib/sdam/topology.js#L737

File: lib/sdam/topology.js (hunk 9 above)
Comment (nbbeeken) [reply lib/sdam/topology.js:255]: Indeed, I'll resolve these

File: lib/utils.js (hunk 14 above)
Comment (nbbeeken) [reply lib/utils.js:669]: Yea JSDoc fixes it, good catch, that is certainly preferable

File: lib/utils.js (hunk 15 above)
Comment (nbbeeken) [reply lib/utils.js:778]: Topology doesn't have the property `ismaster`, Server doesn't have the type `lastIsMaster()` so I have to put type assertions on each access or just here at the top of the function, by "better type" and I should be more explicit we should come up with a better way to assert whether its a Server or Topology.

File: lib/operations/admin_ops.js [hunk 26]
```diff
@@ -59,4 +62,10 @@ function validateCollection(admin, collectionName, options, callback) {
   });
 }
 
+/**
+ * @callback AdminResultCallback
+ * @param {Error} [error]
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply lib/operations/admin_ops.js:67]: I can add for this one, but there's actually 131 things missing documentation, I feel like it would make sense to make another ticket to fill them out, in this epic so its before 4.0 release, does that make sense?
regex for missing doc: `@\w+ \{\w+\} \w+$`

File: lib/sdam/topology.js (hunk 7 above)
Comment (mbroadst) [reply lib/sdam/topology.js:116]: But `CoreCursor` isn't the only one here, `maxPoolSize` and `minHeartbeatFrequencyMS` are both `number`, `reconnect` is a `boolean`, `logger` is a `Logger` instance, etc.

File: lib/cmap/auth/auth_provider.js (hunk 19 above)
[the remark this answers is not in the document]
Comment (nbbeeken) [reply lib/cmap/auth/auth_provider.js:93]: https://github.com/mongodb/node-mongodb-native/pull/2296#discussion_r410218393
